happi/audit.py

```diff
@@ -0,0 +1,424 @@
+
+    def search_pip_package(self, package):
+        """
+        Checks to see if the package is on pypi
+
+        Parameters
+        -----------
+        package: str
+            Name of the package to check
+
+        Returns
+        --------
+            bool
+                To indicate if the package was found or not
+        """
+        process = None
+        try:
+            process = subprocess.Popen('/bin/bash', stdin=subprocess.PIPE,
+                                       stdout=subprocess.PIPE,
+                                       encoding='utf-8')
+            arguments = ['pip', 'search', package]
+            command = ' '.join(arguments)
+
+            out, err = process.communicate(command)
+
+            # the output of 'pip search' will give all the fuzzy searches
+            # we only care if the exact package name is there
+            the_match = re.compile(package)
+            if re.match(the_match, out):
+                return True
+            else:
+                return False
+        except Exception as e:
+            logger.warning(e)
+            if process:
+                process.kill()
+
+    def check_device_in_pypi(self):
+        if self._all_devices:
+            temp_set = self._all_devices.copy()
+            for package in self._all_devices:
+                is_package_found = self.search_pip_package(package)
+                if is_package_found:
+                    logger.info('%s was found in PyPi', package)
+                else:
+                    # maybe a wrong device name was provided because
+                    # it does not exit
+                    logger.info('%s does not exit in PyPi', package)
+                    temp_set.remove(package)
+            # remove the missing package from the set
+            self._all_devices = temp_set
+            return self._all_devices
+
+    def get_value(self, info, item):
+        for key, value in item.items():
+            if key == info.key:
+                return value
+
+    def validate_enforce(self, item):
+        """
+        Validates using enforce_value() from EntryInfo class
+        If the attributes are malformed the entry = contaienr(**item)
+        will fail, thus the enforce_value() will only apply to the
+        ones that were successfully initialized
+        """
+        container = registry[item.get('type')]
+        name = item.get('name')
+        if container:
+            for info in container.entry_info:
+                try:
+                    info.enforce_value(self.get_value(info, item))
+                    return self.report_code.SUCCESS
+                except Exception as e:
+                    logger.info('Invalid value %s, %s', info, e)
+                    return self.report_code.INVALID
+        else:
+            logger.warning('Item %s is missing the container', name)
+            return self.report_code.MISSING
+
+    def print_report_message(self, message):
```

Comment:
A clue something doesn't belong as a method: this doesn't use `self` anywhere.

Reply:
ok I took it outside the class.... thanks :)